core/server/src/main/java/alluxio/worker/file/UnderFileSystemManager.java

```diff
@@ -0,0 +1,170 @@
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.worker.file;
+
+import alluxio.Configuration;
+import alluxio.exception.ExceptionMessage;
+import alluxio.exception.FileAlreadyExistsException;
+import alluxio.exception.FileDoesNotExistException;
+import alluxio.underfs.UnderFileSystem;
+import alluxio.util.IdUtils;
+import alluxio.util.io.PathUtils;
+import alluxio.worker.WorkerContext;
+
+import com.google.common.base.Preconditions;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicLong;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Handles writes to the under file system. Manages open output streams to Under File Systems.
+ * Individual streams should only be used by one client. Each instance keeps its own state of
+ * file names to open streams.
+ */
+@ThreadSafe
+public final class UnderFileSystemManager {
+  /**
+   * A wrapper around the output stream to the under file system. This class handles writing the
+   * data to a temporary file. When the stream is closed, the temporary file will attempt to be
+   * renamed to the final file path. This stream guarantees the temporary file will be cleaned up
+   * when close or cancel is called.
+   */
+  // TODO(calvin): This can be defined by the UnderFileSystem
+  private final class UnderFileSystemOutputStream {
+    /** Configuration to use for this stream. */
+    private final Configuration mConf;
+    /** Underlying stream to the under file system file. */
+    private final OutputStream mStream;
+    /** Final path to write to in the under file system. */
+    private final String mPath;
+    /** Temporary path to write to in the under file system. */
+    private final String mTemporaryPath;
+
+    private UnderFileSystemOutputStream(String ufsPath, Configuration conf)
+        throws FileAlreadyExistsException, IOException {
+      mConf = Preconditions.checkNotNull(conf);
+      mPath = Preconditions.checkNotNull(ufsPath);
+      mTemporaryPath = PathUtils.temporaryFileName(IdUtils.getRandomNonNegativeLong(), mPath);
+      UnderFileSystem ufs = UnderFileSystem.get(mPath, mConf);
+      if (ufs.exists(mPath)) {
+        throw new FileAlreadyExistsException(ExceptionMessage.FAILED_UFS_CREATE.getMessage(mPath));
+      }
+      mStream = ufs.create(mTemporaryPath);
+    }
+
+    /**
+     * Closes the temporary file and deletes it. This object should be discarded after this call.
+     *
+     * @throws IOException if an error occurs during the under file system operation
+     */
+    private void cancel() throws IOException {
```

Comment:
Not sure why we need to implement `cancel` `complete` and `complete` in this wrapper class.  Can't we just make this class a struct of `conf, stream, tmp path, path and etc`, used in the id -> stream map; and consolidate `UnderFileSystemOutputStream.cancel()` with  `UnderFileSystemManager.cancelFile()`?